controllers/users_config_map_controller.go

```diff
@@ -0,0 +1,303 @@
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "failed to get config map", l.Action, l.ActionView)
+		return ctrl.Result{}, err
+	}
+
+	if !controllerutil.ContainsFinalizer(configMap, ConfigMapFinalizer) {
+		controllerutil.AddFinalizer(configMap, ConfigMapFinalizer)
+
+		if err := r.Client.Update(ctx, configMap); err != nil {
+			log.Error(err, "failed to add finalizer to config map", "ConfigMapName", additionalUsersConfigMapName, "Finalizer", ConfigMapFinalizer, l.Action, l.ActionUpdate)
+			return ctrl.Result{}, err
+		}
+		log.Info("finalizer added to configmap")
+		// leave all other steps to reconcile from update action
+		return ctrl.Result{}, nil
+	}
+
+	teamName := getQuayTeamName(configMap.Namespace)
+	removeTeam := false
+
+	if !configMap.DeletionTimestamp.IsZero() {
+		removeTeam = true
+		log.Info("Config map with additional users was removed, will delete team", "TeamName", teamName)
+	}
+
+	var additionalUsers []string
+	if !removeTeam {
+		// get additional users from config map
+		additionalUsersStr, usersExist := configMap.Data[additionalUsersConfigMapKey]
+		if !usersExist {
+			log.Info("Config map with additional users doesn't have the key", "ConfigMapName", additionalUsersConfigMapName, "ConfigMapKey", additionalUsersConfigMapKey, l.Action, l.ActionView)
+			removeTeam = true
+		} else {
+			additionalUsers = strings.Fields(strings.TrimSpace(additionalUsersStr))
+			log.Info("Additional users configured in config map", "AdditionalUsers", additionalUsers)
+		}
+	}
+
+	// reread quay token
+	r.QuayClient = r.BuildQuayClient(log)
+
+	// remove team if config map was removed, or doesn't contain any users
+	if removeTeam {
+		log.Info("Will remove team", "TeamName", teamName)
+		if err := r.QuayClient.DeleteTeam(r.QuayOrganization, teamName); err != nil {
+			log.Error(err, "failed to remove team", "TeamName", teamName, l.Action, l.ActionDelete)
+			return ctrl.Result{}, err
+		}
+
+		// remove finalizer after team is removed
+		if !configMap.DeletionTimestamp.IsZero() {
+			if controllerutil.ContainsFinalizer(configMap, ConfigMapFinalizer) {
+				controllerutil.RemoveFinalizer(configMap, ConfigMapFinalizer)
+				if err := r.Client.Update(ctx, configMap); err != nil {
+					log.Error(err, "failed to remove config map finalizer", "ConfigMapName", additionalUsersConfigMapName, "Finalizer", ConfigMapFinalizer, l.Action, l.ActionUpdate)
+					return ctrl.Result{}, err
+				}
+				log.Info("finalizer removed from config map", l.Action, l.ActionDelete)
+			}
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// fetch image repositories in the namespace
+	imageRepositoryList := &imagerepositoryv1alpha1.ImageRepositoryList{}
+	if err := r.Client.List(ctx, imageRepositoryList, &client.ListOptions{Namespace: configMap.Namespace}); err != nil {
+		log.Error(err, "failed to list ImageRepositories", l.Action, l.ActionView)
+		return ctrl.Result{}, err
+	}
+
+	// get image repositories names
+	repositoriesToAdd := []string{}
```

Comment:
The name is really confusing. In fact that's existing repositories, not that that should be added to the team

Reply:
renamed to "potentialRepositoriesToAdd"

Reply:
While "potentialRepositoriesToAdd" describes your intention, it's still confusing, because the list is actually list of all image repositories in the namespace. Why not to name it like that?

Then `reposToUpdate := filterListDifference(potentialRepositoriesToAdd, reposTeamPermissions)` could became `imageReposToAddToTeam := filterListDifference(allImageRepos, imageReposTeamHasPermissions)`

Reply:
changed